File: desktop/sdk-macos/Sources/EatPassDesktop/EatPassDesktopClient.swift
import DeviceCheck
import EatPassMobileRust
import Foundation

/// Coupled mint: macOS App Attest + attester + issuer + eat-pass finalize.
public final class EatPassDesktopClient: @unchecked Sendable {
    private let config: EatPassConfig
    private let appAttest: DCAppAttestService
    private let keyId: String
    private let credentialPublicKeyHex: String

    public init(
        config: EatPassConfig,
        appAttestService: DCAppAttestService = DCAppAttestService.shared,
        keyId: String,
        credentialPublicKeyHex: String
    ) {
        self.config = config
        self.appAttest = appAttestService
        self.keyId = keyId
        self.credentialPublicKeyHex = credentialPublicKeyHex
    }

    public func mintAuthorizationHeader() async throws -> MintResult {
        let issuerBase = config.issuerUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let attesterBase = config.attesterUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))

        let keysJson = try await httpGet("\(issuerBase)/keys")
        if let pin = config.ktLogPubHex {
            let ktJson = try await httpGet("\(issuerBase)/kt")
            let kt = try JSONSerialization.jsonObject(with: Data(ktJson.utf8)) as? [String: Any]
            let served = (kt?["log_pub"] as? String) ?? ""
            if served.lowercased() != pin.lowercased() {
                throw EatPassError.http("issuer KT log pubkey does not match pinned key")
            }
        }

        let crypto = try EatPassClient(
            issuerPkJson: keysJson,
            issuerName: config.issuerName,
            originInfo: config.originInfo
        )
        let begin = try crypto.begin(count: 1)
        let bundle = try await MacOsAppAttest.createBundle(
            service: appAttest,
            keyId: keyId,
            credentialPublicKeyHex: credentialPublicKeyHex,
            teamId: config.teamId,
            bundleId: config.bundleId,
            bindingHex: begin.bindingHex
        )
        let eatB64 = bundle.base64EncodedString()

        let authBody: [String: Any] = [
            "eat_b64": eatB64,
            "binding": begin.bindingHex,
            "max_batch": 1,
        ]
        let authResp = try await httpPost("\(attesterBase)/authorize", json: authBody)
        guard let authObj = try JSONSerialization.jsonObject(with: Data(authResp.utf8)) as? [String: Any],
              let authorizationB64 = authObj["authorization_b64"] as? String
        else {
            throw EatPassError.http("invalid /authorize response")
        }

        let reqObj = try JSONSerialization.jsonObject(with: Data(begin.requestJson.utf8))
        let signBody: [String: Any] = [
            "req": reqObj,
            "authorization_b64": authorizationB64,
        ]
        let signResp = try await httpPost("\(issuerBase)/sign", json: signBody)
        let headers = try crypto.finalize(signResponseJson: signResp)
        guard let first = headers.first else {
            throw EatPassError.crypto("issuer returned no token")
        }
        return MintResult(authorizationHeader: first, bindingHex: begin.bindingHex)
    }

    private func httpGet(_ url: String) async throws -> String {
        var req = URLRequest(url: URL(string: url)!)
        req.httpMethod = "GET"
        req.timeoutInterval = config.timeoutSeconds
        let (data, resp) = try await URLSession.shared.data(for: req)
        try throwIfHttpError(resp: resp, data: data, url: url)
        return String(decoding: data, as: UTF8.self)
    }

    private func httpPost(_ url: String, json: [String: Any]) async throws -> String {
        var req = URLRequest(url: URL(string: url)!)
        req.httpMethod = "POST"
        req.timeoutInterval = config.timeoutSeconds
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: json)
        let (data, resp) = try await URLSession.shared.data(for: req)
        try throwIfHttpError(resp: resp, data: data, url: url)
        return String(decoding: data, as: UTF8.self)
    }

    private func throwIfHttpError(resp: URLResponse, data: Data, url: String) throws {
        guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let code = (resp as? HTTPURLResponse)?.statusCode ?? -1
            let body = String(decoding: data, as: UTF8.self)
            throw EatPassError.http("HTTP \(code) for \(url): \(body)")
        }
    }
}

File: desktop/sdk-windows/EatPass.Desktop/EatPassDesktopClient.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace EatPass.Desktop;

/// <summary>
/// Coupled mint for Windows agents: TPM attestation + attester + issuer + PoMFRIT finalize.
/// </summary>
public sealed class EatPassDesktopClient : IDisposable
{
    private readonly EatPassConfig _config;
    private readonly HttpClient _http;
    private readonly EatPassNative _native;
    private readonly bool _ownsNative;

    public EatPassDesktopClient(EatPassConfig config, HttpClient? http = null, EatPassNative? native = null)
    {
        _config = config;
        _http = http ?? new HttpClient { Timeout = config.Timeout };
        _native = native ?? new EatPassNative();
        _ownsNative = native is null;
    }

    public void Dispose()
    {
        _http.Dispose();
        if (_ownsNative)
        {
            _native.Dispose();
        }
    }

    public async Task<MintResult> MintAuthorizationHeaderAsync(CancellationToken ct = default)
    {
        var issuerBase = _config.IssuerUrl.TrimEnd('/');
        var attesterBase = _config.AttesterUrl.TrimEnd('/');

        var keysJson = await GetStringAsync($"{issuerBase}/keys", ct);
        if (_config.KtLogPubHex is { } pin)
        {
            var kt = await _http.GetFromJsonAsync<JsonElement>($"{issuerBase}/kt", ct);
            var served = kt.GetProperty("log_pub").GetString() ?? "";
            if (!string.Equals(served, pin, StringComparison.OrdinalIgnoreCase))
            {
                throw new EatPassException("issuer KT log pubkey does not match pinned key");
            }
        }

        using var client = _native.CreateClient(keysJson, _config.IssuerName, _config.OriginInfo);
        try
        {
            var begin = client.Begin(1);
        var bundleJson = await TpmAttestation.CollectBundleJsonAsync(_config, begin.BindingHex, ct);
        var eatB64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(bundleJson));

        var authBody = new { eat_b64 = eatB64, binding = begin.BindingHex, max_batch = 1 };
        using var authResp = await _http.PostAsJsonAsync($"{attesterBase}/authorize", authBody, ct);
        authResp.EnsureSuccessStatusCode();
        var authJson = await authResp.Content.ReadFromJsonAsync<JsonElement>(ct);
        var authorizationB64 = authJson.GetProperty("authorization_b64").GetString()
            ?? throw new EatPassException("missing authorization_b64");

        var signBody = new
        {
            req = JsonSerializer.Deserialize<JsonElement>(begin.RequestJson),
            authorization_b64 = authorizationB64,
        };
        using var signResp = await _http.PostAsJsonAsync($"{issuerBase}/sign", signBody, ct);
        signResp.EnsureSuccessStatusCode();
        var signJson = await signResp.Content.ReadAsStringAsync(ct);
        var header = client.Finalize(signJson);
        return new MintResult(header, begin.BindingHex);
        }
        finally
        {
            client.Dispose();
        }
    }

    private async Task<string> GetStringAsync(string url, CancellationToken ct)
    {
        using var resp = await _http.GetAsync(url, ct);
        resp.EnsureSuccessStatusCode();
        return await resp.Content.ReadAsStringAsync(ct);
    }
}
